File: leetcode/JavaScript实现/100-199题/0144_二叉树的前序遍历.js
// 1、递归 最基本的
var preorderTraversal = function(root) {
    let ans = [];

    let preOrder = (temp, arr) => {
        if (temp == null)
            return;

        arr.push(temp.val);
        preOrder(temp.left, arr);
        preOrder(temp.right, arr);

        return arr;
    };

    preOrder(root, ans);

    return ans;
};

// 2、迭代 考研也经常写非递归的
var preorderTraversal = function(root) {
    let ans = [];
    let stack = [];

    if (!root)
        return [];
    stack.push(root);

    while (stack.length) {
        let temp = stack.pop();
        ans.push(temp.val);

        if (temp.right)
            stack.push(temp.right);
        if (temp.left)
            stack.push(temp.left);
        /*
        这两条语句是上面额if 语句意思一样
        temp.right && stack.push(temp.right);
        temp.left && stack.push(temp.left);
        */
    }

    return ans;
};

File: leetcode/C++实现/100-199题/0144_二叉树的前序遍历.cpp
// 自己熟悉的写法， 也是王道书上的写法
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        stack<TreeNode*> s;
        TreeNode *p = root;

        while (p || !s.empty()) {
            if (p) {
                ans.push_back(p->val);
                s.push(p);
                p = p->left;
            } else {
                TreeNode *temp = s.top();
                s.pop();
                p = temp->right;
            }
        }

        return ans;
    }
};

// 参考这个 统一写法
// https://leetcode-cn.com/problems/n-ary-tree-preorder-traversal/solution/che-di-chi-tou-qian-zhong-hou-xu-di-gui-fa-di-gui-/
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        stack<TreeNode*> s;

        if (root)
            s.push(root);
        
        while (!s.empty()) {
            TreeNode* temp = s.top();
            
            if (temp) {
                s.pop();
                
                if (temp->right)    // 右
                    s.push(temp->right);
                if (temp->left)     // 左
                    s.push(temp->left);

                s.push(temp);       // 中
                s.push(NULL);
            } else {
                s.pop();
                temp = s.top();
                s.pop();
                ans.push_back(temp->val);
            }
        }

        return ans;
    }
};